stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -0,0 +1,162 @@
+        : vari(0.0),
+          M_(A.rows()),
+          _variRefA(reinterpret_cast<vari**>
+                    (stan::math::ChainableStack::memalloc_
+                     .alloc(sizeof(vari*) * A.rows() * (A.cols() + 1) / 2))),
+          _variRefL(reinterpret_cast<vari**>
+                    (stan::math::ChainableStack::memalloc_
+                     .alloc(sizeof(vari*) * A.rows() * (A.rows() + 1) / 2))),
+          _dummy(reinterpret_cast<vari*>(stan::math::ChainableStack::memalloc_
+                     .alloc(sizeof(vari)))) {
+        size_t pos = 0;
+        for (size_type i = 0; i < M_; ++i) {
+          for (size_type j = 0; j <= i; ++j) {
+            _variRefA[pos] = A.coeffRef(i, j).vi_;
+            _variRefL[pos] = new vari(L_A.coeffRef(i, j), false);
+            ++pos;
+          }
+        }
+        _dummy = new vari(0.0, false);
+      }
+
+      /* Reverse mode differentiation 
+       * algorithm refernce: 
+       *
+       * Mike Giles. An extended collection of matrix 
+       * derivative results for forward and reverse mode AD. 
+       * Jan. 2008.
+       *
+       * Note algorithm  as laid out in Giles is 
+       * row-major, so Eigen::Matrices are explicitly storage
+       * order RowMajor, whereas Eigen defaults to 
+       * ColumnMajor. Also note algorithm 
+       * starts by calculating the adjoint for 
+       * A(M_ - 1, M_ - 1), hence pos on line 94 is decremented
+       * to start at pos = M_ * (M_ + 1) / 2.
+       * */
+      virtual void chain() {
+        using Eigen::Matrix;
+        using Eigen::Dynamic;
+        using Eigen::RowMajor;
+        Matrix<double, Dynamic, Dynamic, RowMajor> adjL(M_, M_);
+        Matrix<double, Dynamic, Dynamic, RowMajor> LA(M_, M_);
+        Matrix<double, Dynamic, Dynamic, RowMajor> adjA(M_, M_);
+        size_t pos = 0;
+        for (size_type i = 0; i < M_; ++i)
+          for (size_type j = 0; j <= i; ++j) {
+            adjL.coeffRef(i, j) = _variRefL[pos]->adj_;
+            LA.coeffRef(i, j) = _variRefL[pos]->val_;
+            ++pos;
+          }
+
+        --pos;
+        for (int i = M_ - 1; i >= 0; --i) {
+          for (int j = i; j >= 0; --j) {
+            if (i == j) {
+              adjA.coeffRef(i, j) = 0.5 * adjL.coeffRef(i, j)
+                / LA.coeffRef(i, j);
+            } else {
+              adjA.coeffRef(i, j) = adjL.coeffRef(i, j)
+                / LA.coeffRef(j, j);
+              adjL.coeffRef(j, j) = adjL.coeffRef(j, j)
+                - adjL.coeffRef(i, j) * LA.coeffRef(i, j) / LA.coeffRef(j, j);
+            }
+            for (int k = j - 1; k >=0; --k) {
+              adjL.coeffRef(i, k) = adjL.coeffRef(i, k)
+                - adjA.coeffRef(i, j) * LA.coeffRef(j, k);
+              adjL.coeffRef(j, k) = adjL.coeffRef(j, k)
+                - adjA.coeffRef(i, j) * LA.coeffRef(i, k);
+            }
+            _variRefA[pos--]->adj_ += adjA.coeffRef(i, j);
```

Comment:
I think it's usually more efficient to loop forward than backward.  It's certainly less confusing. Is there any reason this counts down rather than up?